text/00110-compatibility-strategy.md

```diff
@@ -0,0 +1,364 @@
+
+*We want to implement this RFC so that we are able to iterate and change the versioning mechanism itself* 
+
+Actually, We have already acknowledged the need for this a while back:
+
+- https://github.com/aws/cdk-ops/issues/365
+- https://github.com/aws/cdk-ops/issues/233
+
+Its time we solve this once and for all.
+
+After implementing this RFC, we can expect the following outcomes:
+
+- An incompatible change in the CLI won't make it to the release.
+- An incompatible change in the framework, will always be accompanied with a version bump.
+- An incompatible change in the framework, accompanied with a protocol version bump, will always produce the correct error message.
+
+# Design Summary
+
+In order to ensure compatibility between the CLI and the Framework, we have to actually run the CLI against an old framework version, and vice versa.
+We want to find a generic mechanism for running these tests, without having to cherry-pick use-cases or versions. 
+
+The main design idea is to introduce new types of integration tests into our pre-release suite. Those tests will run various permutations of `(CLI X Framework)`, and validate the expected result in each permutation.
+
+To make it generic, the tests will run the entire current integration tests suite, and simply inject different versions each time.
+
+An important note is that when testing compatibility to older versions (be it CLI or Framework), we cannot run the *latest* integration test suite, as it may contain tests for new features and bug fixes, which were not implement in older versions.
+
+Therefore, the entire design relies on us having the ability to:
+
+- *Run **old** integration tests using a **new** CLI version and an **old** Framework version.*
+- *Run **old** integration tests using a **new** Framework version and an **old** CLI version.*
+- *Mark a Framework version as incompatible*
+
+    - The tests should support an ability to specify versions who are not compatible, since the expected output would be different.
+
+# Detailed Design
+
+As mentioned, everything is based on our ability to run integration tests with various version permutations. We introduce a new function, `run_integration_tests_perm` which magically does what we need, as well as some useful constants.
+
+The design is best explained by simply looking at the proposed code changes. Most of what you need to know is in the inline comments of the code.
+
+**[common.bash](https://github.com/aws/aws-cdk/blob/master/packages/aws-cdk/test/integ/cli/common.bash)**
+
+```bash
+function bump_minor() {
+
+    version=$1
+
+    # Accept a version and bump the minor version.
+}
+
+function decrease_minor() {
+
+    version=$1
+
+    # Accept a version and decrease the minor version.
+
+}
+
+# This is the latest published version.
+CURRENT_VERSION=$(node -e 'console.log(require("package.json").version')
```

Comment:
ProTip:

```suggestion
CURRENT_VERSION=$(node -p 'require("package.json").version')
```

Clear up-side is you can't forget to close that parentheses you forgot to close here ... since it no longer exists.

Reply:
The comment on top of this says this is the *latest published version*, but I believe it to be the *soon to be published version*.

Reply:
> $(node -p 'require("package.json").version')

Nice!

And yeah, i fixed the comments.